jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java

```diff
@@ -46,83 +49,97 @@
 
     public GZIPContentDecoder()
     {
-        this(null,2048);
+        this(null, 2048);
     }
 
     public GZIPContentDecoder(int bufferSize)
     {
-        this(null,bufferSize);
+        this(null, bufferSize);
     }
-    
+
     public GZIPContentDecoder(ByteBufferPool pool, int bufferSize)
     {
         _bufferSize = bufferSize;
         _pool = pool;
         reset();
     }
 
-    /** Inflate compressed data from a buffer.
-     * 
-     * @param compressed Buffer containing compressed data.
-     * @return Buffer containing inflated data.
+    /**
+     * Inflates compressed data from a buffer.
+     *
+     * @param compressed the buffer containing compressed data.
+     * @return a buffer containing inflated data.
      */
     public ByteBuffer decode(ByteBuffer compressed)
     {
         decodeChunks(compressed);
-        if (BufferUtil.isEmpty(_inflated) || _state==State.CRC || _state==State.ISIZE )
-            return BufferUtil.EMPTY_BUFFER;
-        
-        ByteBuffer result = _inflated;
-        _inflated = null;
-        return result;
+
+        if (_inflateds.isEmpty())
+        {
+            if (BufferUtil.isEmpty(_inflated) || _state == State.CRC || _state == State.ISIZE)
+                return BufferUtil.EMPTY_BUFFER;
+            ByteBuffer result = _inflated;
+            _inflated = null;
+            return result;
+        }
+        else
+        {
+            _inflateds.add(_inflated);
+            _inflated = null;
+            int length = _inflateds.stream().mapToInt(Buffer::remaining).sum();
+            ByteBuffer result = acquire(length);
+            for (ByteBuffer buffer : _inflateds)
+            {
+                BufferUtil.append(result, buffer);
+                release(buffer);
+            }
+            _inflateds.clear();
+            return result;
+        }
     }
 
-    /** Called when a chunk of data is inflated.
-     * <p>The default implementation aggregates all the chunks 
+    /**
+     * <p>Called when a chunk of data is inflated.</p>
```

Comment:
The javadoc needs a bit of work.  It should mention that it may queue the chunks and that derived implementations should call `release(chunk)` once they have consumed the data.

Reply:
I also don't really understand what happens if true is returned from this method? How does inflating start again?

Reply:
@gregw it's all explained in the javadocs below, apart the `release()` bit.